persistent/Database/Persist/Sql/Types/Internal.hs

```diff
@@ -87,7 +87,7 @@ data SqlBackend = SqlBackend
     -- ^ SQL for inserting many rows and returning their primary keys, for
     -- backends that support this functioanlity. If 'Nothing', rows will be
     -- inserted one-at-a-time using 'connInsertSql'.
-    , connUpsertSql :: Maybe (EntityDef -> NonEmpty UniqueDef -> Text -> Text)
+    , connUpsertSql :: Maybe (EntityDef -> NonEmpty (HaskellName,DBName) -> Text -> Text)
```

Comment:
I think it's necessary, because I don't see a way to go from `Unique record` (which is the type of `uniqueKey` in `Persist/Sql/Orphan/PersistUnique.hs`) to its corresponding `UniqueDef`.

I'd be glad to be proven wrong.

Reply:
Hmmm. That's irritating.

Looks like we have this possible pathway:

1. With a `PersistEntity e` in scope, we can call `let uniqDefs = entityUniques (entityDef (Proxy @e)) :: [UniqueDef]`.
2. Then we have `persistUniqueToFieldNames :: Unique record -> [(HaskellName, DBName)]`.
3. We can do a lookup on the `[UniqueDef]` to find the corresponding `uniqueFields` - eg `find (\u -> uniqueFields u == persistUniqueToFieldNames uniq) uniqDefs`

Not the most elegant solution, for sure, but it will work I think.